### PR TITLE
brew.rb: only print "Kernel.exit" on failures.

### LINKED ISSUE #47099: Executing external commands: "Kernel.exit" is always printed to stdout when --verbose is set, even if exit status is 0
Consider a very simple external command `meh`, which only consists of `puts "meh"`:

```zsh
> echo 'puts "meh"' >brew-meh.rb
> chmod +x brew-meh.rb
> PATH=$PWD:$PATH
> brew meh --verbose 2>/dev/null
meh
Kernel.exit
```

As you can see, `Kernel.exit` is printed on stdout although there's no error whatsoever, which is rather surprising and obnoxious. The rightful place for such a message, IMO, is stderr, and ideally it shouldn't appear at all when the exit status is 0.

This issue seems to be due to https://github.com/Homebrew/homebrew/blob/80bdf34/Library/brew.rb#L150 always `exit`ing after requiring the external command, which is then picked up by `rescue SystemExit` https://github.com/Homebrew/homebrew/blob/80bdf34/Library/brew.rb#L178.

This issue is also causing some minor regression for homebrew-cask after adopting the external command architecture. See https://github.com/caskroom/homebrew-cask/issues/15882.

### PR DESCRIPTION
**Pull request by mikemcquaid**
CC @zmwangx

Closes #47099.

### CODE REVIEW

File: Library/brew.rb [hunk 1]
```diff
@@ -175,8 +175,8 @@ def require?(path)
 rescue UsageError
   onoe "Invalid usage"
   abort ARGV.usage
-rescue SystemExit
-  puts "Kernel.exit" if ARGV.verbose?
+rescue SystemExit => e
+  puts "Kernel.exit" if ARGV.verbose? && !e.success?
```
Comment (UniqMartin) [inline Library/brew.rb]: Should we make that `$stderr.puts` instead? Doesn't seem helpful to mix this with regular output.

**Comment by UniqMartin** [comment]
Just a very minor suggestion, otherwise: :+1:

**Comment by zmwangx** [comment]
:+1: I'd also like to have this message on stderr, by the way.

**Comment by DomT4** [comment]
Fine with me.

**Comment by mikemcquaid** [comment]
Updated to use `onoe` (for `STDERR`) and optionally print the backtrace when in debug mode.

**Comment by UniqMartin** [comment]
> Updated to use `onoe` (for `STDERR`) […]

Not sure if we really want the formatted `Error:` prefix also printed by `onoe`, but I don't mind.

> […] and optionally print the backtrace when in debug mode.

:+1:

**Comment by mikemcquaid** [comment]
> Not sure if we really want the formatted Error: prefix also printed by onoe, but I don't mind.

Because I'm checking the return code: it will always be an error in that case.

**Comment by UniqMartin** [comment]
> Because I'm checking the return code: it will always be an error in that case.

Yes, I realized that, but thanks for clarifying. My question was purely about aesthetics, i.e., not important.

**Pull request closed without merging**